### PR TITLE
HDDS-9201. [Ozone-Streaming] Stream copy object support on gateway

### PR DESCRIPTION
**Pull request by guohao-rosicky**
## What changes were proposed in this pull request?

Stream copy object support on gateway.

The current gateway only supports uploading files using craft in the copy object section, this pr uses stream to support the copy object.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-9201

## How was this patch tested?

Existing UT

### CODE REVIEW

**Comment by guohao-rosicky** [comment]
all ci passed, see: https://github.com/guohao-rosicky/ozone/actions/runs/5946593623

File: hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/ObjectEndpoint.java [hunk 1]
```diff
@@ -1004,11 +1005,20 @@ void copy(OzoneVolume volume, InputStream src, long srcKeyLen,
       ReplicationConfig replication,
             Map<String, String> metadata) throws IOException {
     long copyLength;
-    try (OzoneOutputStream dest =
-                 getClientProtocol().createKey(
-        volume.getName(), destBucket, destKey, srcKeyLen,
-        replication, metadata)) {
-      copyLength = IOUtils.copyLarge(src, dest);
+    if (datastreamEnabled && !((replication != null &&
```
Comment (hemantk-12) [inline hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/ObjectEndpoint.java:1008]:
```suggestion
    if (datastreamEnabled && !(replication != null &&
        replication.getReplicationType() == EC) &&
        srcKeyLen > datastreamMinLength) {
```

Please remove redundant parentheses.

**Review by hemantk-12 — COMMENTED** [review]
Thanks @guohao-rosicky for working on this.

1. PR/Jira description is missing the motivation behind the change. It would be great if you add more details there.
2. I doubt if existing test are sufficient. We might not be reaching here: https://github.com/apache/ozone/pull/5210/files#diff-bd6abd262ed8e24eaeea5cb0e7c2a1a128411b284f9f27dae1ad47be6717a940R1011 based on current tests. Please correct me if we already have test for it.

**Comment by guohao-rosicky** [comment]
Thanks @hemantk-12 for the review, all ci is passed, please see: https://github.com/guohao-rosicky/ozone/actions/runs/6019656855

**Comment by guohao-rosicky** [comment]
> 1. PR/Jira description is missing the motivation behind the change. It would be great if you add more details there.
hi, @hemantk-12, The description has been updated

**Review by hemantk-12 — COMMENTED** [review]
LGTM.

**Review by captainzmc — APPROVED** [review]
The change looks good. Wait for CI execution.

**Comment by guohao-rosicky** [comment]
Thanks @captainzmc for the review, the ci it looks like the rename error in filesystem has nothing to do with this commit.

**Comment by captainzmc** [comment]
Thanks @guohao-rosicky for the patch. And thanks @hemantk-12 for the review. Let's merge this.

**Pull request merged**